File: Python/LRU Cache.py
class LRUCache(object):
    class ListNode(object):
        def __init__(self, key, value):
            self.key = key
            self.value = value
            self.next = None
            self.pre = None

    """
    :type capacity: int
    """
    def __init__(self, capacity):
        self.cap = capacity
        self.head = self.ListNode(-1,-1)
        self.tail = self.ListNode(-1,-1)
        self.head.next = self.tail
        self.tail.pre = self.head
        self.map = {}

    """
    :rtype: int
    """
    def get(self, key):
        if key in self.map:
            node = self.map[key]
            self.remove(node)
            self.addFirst(node)
            return node.value
        else:
            return -1

    """
    :type key: int
    :type value: int
    :rtype: nothing
    """
    def set(self, key, value):
        if key in self.map:
            node = self.map[key]
            node.value = value
            self.remove(node)
            self.addFirst(node)
        else:
            if len(self.map) >= self.cap:
                node = self.tail.pre
                self.remove(node)
                del self.map[node.key]
            node = self.ListNode(key, value)
            self.addFirst(node)
            self.map[key] = node

    def remove(self, node):
        node.pre.next = node.next
        node.next.pre = node.pre

    def addFirst(self, node):
        node.pre = self.head
        node.next = self.head.next
        self.head.next = node
        node.next.pre = node

File: Java/LRU Cache.java
public class LRUCache {
    private class ListNode {
        ListNode pre;
        ListNode next;
        int key;
        int value;
        
        public ListNode(int key, int value) {
            this.key   = key;
            this.value = value;
        }
    }
    
    private int cap = 0;
    private HashMap<Integer, ListNode> map = new HashMap<Integer, ListNode>();
    private ListNode head = new ListNode(-1, -1);
    private ListNode tail = new ListNode(-1, -1);
    
    
    public LRUCache(int capacity) {
        cap = capacity;
        head.next = tail;
        tail.pre = head;
    }
    
    public int get(int key) {
        if (map.containsKey(key)) {
            ListNode node = map.get(key);
            remove(node);
            addFirst(node);
            return node.value;
        }
        return -1;
    }
    
    public void set(int key, int value) {
        if (map.containsKey(key)) { // set
            ListNode node = map.get(key);
            node.value = value;
            remove(node);
            addFirst(node);
        } else { // add
            if (map.size() >= this.cap) {
                ListNode lastNode = tail.pre;
                remove(lastNode);
                map.remove(lastNode.key);
            }
            ListNode newNode = new ListNode(key, value);
            map.put(key, newNode);
            addFirst(newNode);
        }
    }
    
    private void remove(ListNode node) {
        if (node == null)
            return;
        node.pre.next = node.next;
        node.next.pre = node.pre;
    }
    
    private void addFirst(ListNode node) {
        if (node == null)
            return;
        node.pre = head;
        node.next = head.next;
        node.next.pre = node;
        head.next = node;
    }
}